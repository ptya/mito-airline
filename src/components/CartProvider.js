import React, { createContext, useReducer } from "react";

const CartContext = createContext();

const cartReducer = (state, action) => {
  switch (action.type) {
    case "setInbound":
      return {
        ...state,
        inbound: action.flight,
        inboundPrice: action.price,
        total: state.outboundPrice + action.price
      };
    case "setOutbound":
      return {
        ...state,
        outbound: action.flight,
        outboundPrice: action.price,
        total: state.inboundPrice + action.price
      };
    case "purgeInbound":
      return {
        ...state,
        inbound: null,
        inboundPrice: 0,
        total: state.outboundPrice
      };
    case "purgeOutbound":
      return {
        ...state,
        outbound: null,
        outboundPrice: 0,
        total: state.inboundPrice
      };
    default:
      return state;
  }
};

const initialCart = {
  inbound: null,
  inboundPrice: 0,
  outbound: null,
  outboundPrice: 0,
  total: 0
};

const CartProvider = props => {
  const [cart, cartDispatch] = useReducer(cartReducer, initialCart);

  return (
    <CartContext.Provider
      value={{
        cart,
        cartDispatch
      }}
    >
      {props.children}
    </CartContext.Provider>
  );
};

export default CartProvider;
export { CartContext };
