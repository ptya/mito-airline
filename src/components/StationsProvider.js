import React, { createContext, useState, useEffect } from "react";
import PropTypes from "prop-types";

const StationsContext = createContext();

const originSession = "mito-origin";
const destSession = "mito-destination";

const StationsProvider = props => {
  const [stations, setStations] = useState();
  const [origin, setOrigin] = useState(
    JSON.parse(sessionStorage.getItem(originSession)) || null
  );
  const [destination, setDestination] = useState(
    JSON.parse(sessionStorage.getItem(destSession)) || null
  );
  const [departureDate, setDepartureDate] = useState();
  const [returnDate, setReturnDate] = useState();
  const [secondaryReturnDate, setSecondaryReturnDate] = useState();

  const fetchStations = async () => {
    const res = await fetch("https://mock-air.herokuapp.com/asset/stations");
    const data = await res.json();
    setStations(data);
  };

  useEffect(() => {
    fetchStations();
  }, []); // empty array to run only on mount

  return (
    <StationsContext.Provider
      value={{
        stations,
        origin,
        destination,
        departureDate,
        returnDate,
        secondaryReturnDate,
        setOrigin,
        setDestination,
        setDepartureDate,
        setReturnDate,
        setSecondaryReturnDate
      }}
    >
      {stations ? props.children : null}
    </StationsContext.Provider>
  );
};

StationsProvider.propTypes = {
  children: PropTypes.node.isRequired
};

export default StationsProvider;
export { StationsContext, originSession, destSession };
