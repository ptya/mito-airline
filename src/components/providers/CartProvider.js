import React, { createContext, useReducer } from "react";
import PropTypes from "prop-types";

const CartContext = createContext();

const initialCart = {
  inbound: null,
  inboundPrice: 0,
  inboundFare: "",
  outbound: null,
  outboundPrice: 0,
  outboundFare: "",
  total: 0
};

const cartReducer = (state, action) => {
  switch (action.type) {
    case "setInbound":
      return {
        ...state,
        inbound: action.flight,
        inboundPrice: action.price,
        inboundFare: action.fare,
        total: state.outboundPrice + action.price
      };
    case "setOutbound":
      return {
        ...state,
        outbound: action.flight,
        outboundPrice: action.price,
        outboundFare: action.fare,
        total: state.inboundPrice + action.price
      };
    case "purge":
      return initialCart;
    default:
      return state;
  }
};

const CartProvider = props => {
  const [cart, cartDispatch] = useReducer(cartReducer, initialCart);

  return (
    <CartContext.Provider
      value={{
        cart,
        cartDispatch
      }}
    >
      {props.children}
    </CartContext.Provider>
  );
};

CartProvider.propTypes = {
  children: PropTypes.node.isRequired
};

export default CartProvider;
export { CartContext, cartReducer, initialCart };
