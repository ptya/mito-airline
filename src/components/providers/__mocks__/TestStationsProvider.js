import React, { useReducer } from "react";
import PropTypes from "prop-types";

import {
  StationsContext,
  stationsReducer,
  initialStations
} from "../StationsProvider";

const TestStationsProvider = props => {
  const { state, reducer } = props;
  const [stations, stationsDispatch] = useReducer(reducer, state);

  return (
    <StationsContext.Provider
      value={{
        stations,
        stationsDispatch
      }}
    >
      {props.children}
    </StationsContext.Provider>
  );
};

TestStationsProvider.propTypes = {
  children: PropTypes.node.isRequired,
  state: PropTypes.object,
  reducer: PropTypes.func
};

TestStationsProvider.defaultProps = {
  state: initialStations,
  reducer: stationsReducer
};

export default TestStationsProvider;
