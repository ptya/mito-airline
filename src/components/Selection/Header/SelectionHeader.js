import React, { useContext } from "react";

// context
import { StationsContext } from "components/providers/StationsProvider";
import { CartContext } from "components/providers/CartProvider";

// local elements
import Header from "./elements/Header";
import Logo from "./elements/Logo";
import Route from "./elements/Route";
import Switch from "./elements/Switch";

const SelectionHeader = () => {
  const {
    stationsDispatch,
    stations: { origin, destination }
  } = useContext(StationsContext);

  const { cartDispatch } = useContext(CartContext);

  const handleReturn = () => {
    stationsDispatch({ type: "clearDates" });
  };

  const handleSwitch = () => {
    stationsDispatch({ type: "switchStations" });
    cartDispatch({ type: "purge" });
  };

  return (
    <Header>
      <Logo to="/" handler={handleReturn} alt={"Mito Airline"} />
      <Route>
        <span>Leaving from</span>
        {origin.shortName}
      </Route>
      <Switch handler={handleSwitch} />
      <Route>{destination.shortName}</Route>
    </Header>
  );
};

export default SelectionHeader;
