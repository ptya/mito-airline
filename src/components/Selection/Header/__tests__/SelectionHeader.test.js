import React from "react";
import { render, fireEvent } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";

// components
import SelectionHeader from "../SelectionHeader";

// context
import TestStationsProvider from "components/providers/__mocks__/TestStationsProvider";
import CartProvider from "components/providers/CartProvider";
import { stationsReducer } from "components/providers/StationsProvider";

// mocks
import {
  stationOne,
  stationTwo
} from "components/providers/__mocks__/mockValues";

afterEach(() => {
  console.error.mockClear();
  reducer.mockClear();
});

console.error = jest.fn();
const reducer = jest.fn(stationsReducer);

function RTLrender() {
  const state = {
    origin: stationOne,
    destination: stationTwo
  };
  const context = render(
    <Router>
      <TestStationsProvider state={state} reducer={reducer}>
        <CartProvider>
          <SelectionHeader />
        </CartProvider>
      </TestStationsProvider>
    </Router>
  );
  return context;
}

test("Renders without errors and matches snapshot", () => {
  const { container } = RTLrender();

  expect(container).toMatchSnapshot();
  expect(console.error).not.toHaveBeenCalled();
});

test("Switching between stations updates header", () => {
  const { getByTestId } = RTLrender();

  expect(getByTestId("sh-origin").textContent).toEqual(
    expect.stringContaining(stationOne.shortName)
  );
  expect(getByTestId("sh-dest").textContent).toBe(stationTwo.shortName);

  fireEvent.click(getByTestId("sh-switch"));

  expect(getByTestId("sh-origin").textContent).toEqual(
    expect.stringContaining(stationTwo.shortName)
  );
  expect(getByTestId("sh-dest").textContent).toBe(stationOne.shortName);

  expect(console.error).not.toHaveBeenCalled();
});

test("Clicking on logo redirects to home by clearing dates", () => {
  const { getByTestId } = RTLrender();

  fireEvent.click(getByTestId("sh-logo"));
  expect(reducer).toHaveBeenCalledWith(
    expect.any(Object),
    expect.objectContaining({
      type: "clearDates"
    })
  );

  expect(console.error).not.toHaveBeenCalled();
});
