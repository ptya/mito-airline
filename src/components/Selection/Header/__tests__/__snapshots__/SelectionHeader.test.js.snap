// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders without errors and matches snapshot 1`] = `
.c1 {
  display: inline-block;
  vertical-align: middle;
  border: 0;
  background: #06038D;
  margin: auto 0 auto 15px;
  cursor: pointer;
}

.c1 img {
  height: 29px;
  width: 28px;
  margin: 0;
  -webkit-transform: rotateX(0);
  -ms-transform: rotateX(0);
  transform: rotateX(0);
  -webkit-transition: -webkit-transform 0.5s ease-in-out;
  -webkit-transition: transform 0.5s ease-in-out;
  transition: transform 0.5s ease-in-out;
}

.c1 img:hover {
  -webkit-transform: rotate(360deg);
  -ms-transform: rotate(360deg);
  transform: rotate(360deg);
}

.c3 {
  border: 0;
  background-color: initial;
  cursor: pointer;
}

.c3 img {
  width: 30px;
  margin: auto 0;
}

.c0 {
  background: #06038D;
  color: #FFF;
  z-index: 100;
  position: fixed;
  height: 60px;
  top: 0;
  left: 0;
  right: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0::after {
  content: "";
  width: 100%;
  height: 20px;
  background-image: linear-gradient(#f2f2f2,#f2f2f200);
  position: absolute;
  bottom: -20px;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  font-size: 18px;
  margin: 12px 15px;
  overflow: hidden;
}

.c2 .route-leaving {
  display: block;
  text-transform: uppercase;
  font-size: 10px;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c2 span {
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
}

@media (min-width:650px) {
  .c1 {
    margin: auto 80px auto 20px;
  }
}

@media (min-width:650px) {
  .c2 {
    margin: 12px 40px;
    font-size: 20px;
  }
}

<div>
  <header
    class="c0"
  >
    <a
      class="c1"
      data-testid="sh-logo"
      href="/"
    >
      <img
        alt="Mito Airline"
        src="mito-logo.svg"
      />
    </a>
    <p
      class="c2"
      data-testid="sh-origin"
    >
      <span
        class="route-leaving"
      >
        Leaving from
      </span>
      <span>
        Barcelona
      </span>
    </p>
    <button
      class="c3"
      data-testid="sh-switch"
      style="transform: rotateZ(0deg);"
    >
      <img
        alt="From - to"
        src="arrows.svg"
      />
    </button>
    <p
      class="c2"
      data-testid="sh-dest"
    >
      <span>
        Budapest
      </span>
    </p>
  </header>
</div>
`;
